### DISCUSSION TITLE
How to increase performance

### DISCUSSION THREAD

**Discussion by nath1as** [piroor/treestyletab#3007] [Q&A] [answered]
I am using a keyboard to shift through tabs and would like to make this as fast as possible. I have disabled some delays but still have some noticeable lag. 
Does anyone have any suggestions for setting things to make it more performant?

**Comment by piroor** [comment, marked as answer]
Which delay did you talk about? Detailed description may help us to advice.

**Reply by nath1as** [reply]
switching from the current tab to the next tab has a delay, I would like to minimize it but I'm not sure if its a limitation of firefox or this plugin

after disabling animations I think firefox is the bottleneck

**Reply by piroor** [reply]
Hmm... Let's explain how TST switch active tab on the case.

1. TST's sidebar panel receives `mousedown` event on a sidebar tab.
2. If there is any helper addons registered as `mousedown` listeners, TST sends notification messages to helper addons and wait responses.
3. TST's sidebar panel sends a message activating the tab to the background page.
4. TST's background page receives the message, and calls `tabs.update()` or `tabs.highlight()` to activate the specified tab.

* When Firefox is slow, messaging at steps 2, 3 and 4 may be delayed.
* If you have some helper addons, messaging process at the step 2 may take time. Disabling helper addons and reloading TST may decrease delay from this reason.

If we prefer performance than robustness or features, we should call `tabs.update()` (or `tabs.highlight()`) immediately at the step 1. In other words, TST is fatally slow at this point because I preferred robustness. I think we can reconstruct this mechanism totally for better performance and robustness, but currently I cannot imagine how to do that due to my technical capabilities, sorry.

Anyway patches are welcome. I hope that TST gets better performance safely, so I'll merge pullrequests for that.
